src/structs/partial_buffer.rs

```diff
@@ -0,0 +1,90 @@
+use std::cmp::min;
+
+/// This struct is used to the fact that we have to take buffers
+/// as they arrive, and we might not have all the data we need.
+///
+/// This used is via the take function, which attempts to grab
+/// the amount of data specified, and if it isn't available, stores
+/// the partial data in the extra buffer, and returns None.
+///
+/// Next time around, the extra data will be prepended to the next
+/// buffer, so eventually the amount of data requested will become
+/// available.
+///
+/// The concept of reserve is used to handle the case where we need
+/// to leave a certain amount of data in the buffer, perticularly
+///  where Lepton files have the 32bit file size appended.
+/// We don't want this to get parsed out, so we ensure that there are
+/// always at least 4 bytes in the buffer.
+pub struct PartialBuffer<'a> {
+    slice: &'a [u8],
+    extra_buffer: &'a mut Vec<u8>,
+    continue_processing: bool,
+}
+
+impl<'a> PartialBuffer<'a> {
+    pub fn new(slice: &'a [u8], extra_buffer: &'a mut Vec<u8>) -> PartialBuffer<'a> {
+        PartialBuffer {
+            slice,
+            extra_buffer,
+            continue_processing: true,
+        }
+    }
+
+    pub fn continue_processing(&self) -> bool {
+        self.continue_processing
+    }
+
+    /// grabs a variable amount of data if is available, otherwise put it in the extra
+    /// buffer for next time.
+    pub fn take(&mut self, size: usize, reserve: usize) -> Option<Vec<u8>> {
```

Comment:
So, I understand what "reserve" here means - do not return the data unless the buffers also have addition "reserve" bytes. But it is not self-explaining. Can you think of a better name?